src/torchcodec/decoders/_core/VideoDecoder.cpp

```diff
@@ -497,6 +485,36 @@ void VideoDecoder::addVideoStream(
       formatContext_->streams[i]->discard = AVDISCARD_ALL;
     }
   }
+}
+
+void VideoDecoder::addVideoStream(
+    int streamIndex,
+    const VideoStreamOptions& videoStreamOptions) {
+  TORCH_CHECK(
+      videoStreamOptions.device.type() == torch::kCPU ||
+          videoStreamOptions.device.type() == torch::kCUDA,
+      "Invalid device type: " + videoStreamOptions.device.str());
+
+  addStream(streamIndex, AVMEDIA_TYPE_VIDEO, videoStreamOptions.device);
+
+  auto& streamMetadata =
+      containerMetadata_.allStreamMetadata[activeStreamIndex_];
+
+  if (seekMode_ == SeekMode::approximate &&
+      !streamMetadata.averageFps.has_value()) {
+    throw std::runtime_error(
+        "Seek mode is approximate, but stream " +
+        std::to_string(activeStreamIndex_) +
+        " does not have an average fps in its metadata.");
+  }
+
+  auto& streamInfo = streamInfos_[activeStreamIndex_];
+  streamInfo.videoStreamOptions = videoStreamOptions;
+  streamInfo.codecContext->thread_count =
+      videoStreamOptions.ffmpegThreadCount.value_or(0);
```

Comment:
TODO need to check before merging whether this should be put back within `addStream()`, before the call to `initializeContextOnCuda()`.

Reply:
Going back to this:  `initializeContextOnCuda` doesn't seem to rely on `threadCount` at all, but `avcodec_open2` definitely does.
So I'll have to put this back, which is going to be ugly :/